process/track/baseBlockTrack.go

```diff
@@ -0,0 +1,425 @@
+}
+
+func (bbt *baseBlockTrack) displayTrackedHeadersForShard(shardID uint32, message string) {
+	headers, hashes := bbt.sortHeadersFromNonce(shardID, 0)
+	shouldNotDisplay := len(headers) == 0 ||
+		len(headers) == 1 && headers[0].GetNonce() == 0
+	if shouldNotDisplay {
+		return
+	}
+
+	log.Debug(message,
+		"shard", shardID,
+		"nb", len(headers))
+
+	for index, header := range headers {
+		log.Trace("tracked header info",
+			"round", header.GetRound(),
+			"nonce", header.GetNonce(),
+			"hash", hashes[index])
+	}
+}
+
+// GetCrossNotarizedHeader returns a cross notarized header for a given shard with a given offset, behind last cross notarized header
+func (bbt *baseBlockTrack) GetCrossNotarizedHeader(shardID uint32, offset uint64) (data.HeaderHandler, []byte, error) {
+	return bbt.crossNotarizer.getNotarizedHeader(shardID, offset)
+}
+
+// GetLastCrossNotarizedHeader returns last cross notarized header for a given shard
+func (bbt *baseBlockTrack) GetLastCrossNotarizedHeader(shardID uint32) (data.HeaderHandler, []byte, error) {
+	return bbt.crossNotarizer.getLastNotarizedHeader(shardID)
+}
+
+// GetLastHeader returns the last header received (highest round) for the given shard
+func (bbt *baseBlockTrack) GetLastHeader(shardID uint32) data.HeaderHandler {
+	bbt.mutHeaders.RLock()
+	defer bbt.mutHeaders.RUnlock()
+
+	var lastHeaderForShard data.HeaderHandler
+
+	headersForShard, ok := bbt.headers[shardID]
+	if !ok {
+		return lastHeaderForShard
+	}
+
+	maxRound := uint64(0)
+	for _, headersInfo := range headersForShard {
+		for _, headerInfo := range headersInfo {
+			if headerInfo.header.GetRound() > maxRound {
+				maxRound = headerInfo.header.GetRound()
+				lastHeaderForShard = headerInfo.header
+			}
+		}
+	}
+
+	return lastHeaderForShard
+}
+
+// GetTrackedHeadersForShard returns tracked headers for a given shard
+func (bbt *baseBlockTrack) GetTrackedHeaders(shardID uint32) ([]data.HeaderHandler, [][]byte) {
+	return bbt.sortHeadersFromNonce(shardID, 0)
+}
+
+func (bbt *baseBlockTrack) sortHeadersFromNonce(shardID uint32, nonce uint64) ([]data.HeaderHandler, [][]byte) {
+	bbt.mutHeaders.RLock()
+	defer bbt.mutHeaders.RUnlock()
+
+	headersForShard, ok := bbt.headers[shardID]
+	if !ok {
+		return nil, nil
+	}
+
+	sortedHeadersInfo := make([]*headerInfo, 0)
+
+	for headersNonce, headersInfo := range headersForShard {
+		if headersNonce < nonce {
+			continue
+		}
+
+		sortedHeadersInfo = append(sortedHeadersInfo, headersInfo...)
+	}
+
+	if len(sortedHeadersInfo) > 1 {
+		sort.Slice(sortedHeadersInfo, func(i, j int) bool {
+			return sortedHeadersInfo[i].header.GetNonce() < sortedHeadersInfo[j].header.GetNonce()
+		})
+	}
+
+	headers := make([]data.HeaderHandler, 0)
+	headersHashes := make([][]byte, 0)
+
+	for _, headerInfo := range sortedHeadersInfo {
+		headers = append(headers, headerInfo.header)
+		headersHashes = append(headersHashes, headerInfo.hash)
+	}
+
+	return headers, headersHashes
```

Comment:
these results could be cached, and redo the sorting when changes were done.

Reply:
This method is will be called only one time per round in BlockProcessor and it does only (1-2 appends for each call) as it returns headers which are stored per shard / per nonce. It will have two records when there are forks(network split) so maximum 2 records for the same nonce.